[FLINK-1722][datastream] Enable the InitializeOnMaster and FinalizeOnMaster interfaces on datastream

sunhaibotb opened the pull request:
## What is the purpose of the change

This pull request enables the InitializeOnMaster and FinalizeOnMaster interfaces on datastream. The aim is to merge blink batch tables into master and move toward  the goal of streaming-batch unification.


## Brief change log

  - Introduces a new class `InputOutputFormatVertex` (a derived class of JobVertex) to handle global initialization and finalization of `InputFormat(s)` and `OutputFormat(s)`.
  - Introduces a new class `InputOutputFormatStub` to handle the format stubs of `InputFormat(s)` and `OutputFormat(s)`. 
  - Changes the logic that generates `JobGraph` from `StreamGraph`, and when a job vertex contains `InputFormat` or `OutputFormat`, create it with `InputOutputFormatVertex.`
  - Replaces all places where `InputFormatVertex` and `OutputFormatVertex `are used with `InputOutputFormatVertex`, and delete `InputFormatVertex` and `OutputFormatVertex`.


## Verifying this change

This change added tests and can be verified as follows:

  - Extended test that validates the logic of `InputOutputFormatVertex`.
  - Extended test that checks if `InputOutputFormatVertex` is created correctly when a job contains `InputFormat` and `OutputFormat` on generating a job graph.
  - Added test for new `InputOutputFormatStub` class.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

sunhaibotb commented:
@aljoscha Can you help reviewing this PR? Merging blink tables into master requires this feature. Thank you very much.

aljoscha commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/InputOutputFormatStub.java:
```diff
@@ -0,0 +1,201 @@
+/**
+ * A utility class for {@link InputFormat} and {@link OutputFormat} stubs.
+ */
+public class InputOutputFormatStub {
```
I think we need a better name for this, maybe `InputOutputFormatContainer`.

The word `stub` is very old in the code base and possibly comes from a misunderstanding of the word. I actually don't know. I think nowadays it means placeholder or something like this. 😅

aljoscha commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/InputOutputFormatStub.java:
```diff
@@ -0,0 +1,201 @@
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
```
This could be
```
A container for {@link InputFormat InputFormats} and {@link OutputFormat OutputFormats}, along with their {@link Configuration.
```

aljoscha commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/InputOutputFormatStub.java:
```diff
@@ -0,0 +1,201 @@
+	public InputOutputFormatStub() {
+		this.formats = new FormatUserCodeTable();
+		this.parameters = new Configuration();
+		this.userCodeClassLoader = Thread.currentThread().getContextClassLoader();
```
It might be better to pass this in the constructor, just to be explicit. Feel free to leave it like this, though.

sunhaibotb commented:
Thank you for reviewing @aljoscha . Because `JobMasterTest.Java` conflicts with the latest master, I will rebase on the top of the latest master and force push after addressing the comments.

sunhaibotb commented:
The PR has been updated and all conflicts resolved. @aljoscha

aljoscha commented:
I merged this! Thanks for the good work @sunhaibotb :👌
